Add reverse traversal to iter

icristescu opened the pull request:
Replaced bfs with dfs in `iter` in order to support traversal of the closure graph in both direction: when `rev = false` then a node (and the edges to its predecessors) is treated when it is first visited. Otherwise a node is treated only after all of its predecessors have been treated first. `skip` and `level` are verified when first visiting each node.

samoht commented on src/irmin/irmin.mli:
```diff
@@ -1747,12 +1747,15 @@ module Private : sig
+        ?rev:bool ->
         unit ->
         unit Lwt.t
       (** [iter min max node edge skip ()] is the same as closure except that it
```
Can you also clarify that comment -- `iter` is not exactly like `same as closure`. I think you mean the parameter of the function have the same meaning? I think we should state what this function is doing a bit more clearly. What about:

```
[iter ...] iterates over a subgraph in topological order. If [rev] is false (the default), we have the guarantee that all the predecessors of a node will be traversed before `node` is called.
```
We also need to say which guarantee do we have for `edge`: will it be called once both the source and the destination is traversed, or is it done lazily?

icristescu replied:
I changed it, it might be too much text now. I would like to keep `closure graph` in the doc so as not too specify again the constraints in the doc of `closure` (too be specified [here](https://github.com/mirage/irmin/pull/908) )

samoht commented:
Thanks, that looks great. Just need to explain the semantics a bit better (especially for the `edge` function) and that'll be good to merge!

icristescu commented:
I marked this WIP because I realised there are two differences bwt `iter` and `closure` and that can lead to confusion with the semantics of `iter`:
- the min nodes are not visited in `iter` but they are added to the closure graph. If we don't want to visit the min nodes maybe we can remove them from `iter` (`skip` can be used instead).
- if an edge is included in the closure graph then both its vertices are in the graph as well, while in `iter` we visit an edge even if it points to a skipped node.

icristescu commented:
- Changed `iter` to also visit the `min` nodes. The difference with `skip` is that neither a skipped node nor its predecessors are visited; a `min` node is visited but not its predecessors (or they are but through different paths). Therefore it makes sense to keep both `min` and `skip` as arguments.
- Edges that are visited can point to a skipped node. Apart from this case, both the source and the destination of an edge are visited.
 
I also modified the doc. What a `closure graph between max and min` means is explained in PR #908 . 

If you prefer a different semantics for `min` or `edge`, or a different name for the `skip` and `min` arguments, I can modify this.

samoht commented:
That's a good point that `min` and `skip` are actually pretty similar... do we need both?

icristescu commented:
> That's a good point that `min` and `skip` are actually pretty similar... do we need both?

I don't see a straightforward way to combine them. `skip` checks nodes as it traverses the graph, while `min` has to be specified before starting the traversal.  Also, because the min nodes have to be visited (for building the closure graph) they can't be specified using `skip`.
